Review followup: PR #23022 — GH#23019: enforce ruleset required checks in pulse merge gate

## Unaddressed review bot suggestions

PR #23022 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/23022

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `AGENTS.md`
"AI-Generated Issue Quality"). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `AGENTS.md` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- _(No file paths in inline comments — see PR review summaries below for context)_

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

_(none)_

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/23022#pullrequestreview-4237910589)

> ## Code Review
> 
> This pull request enhances the pulse-merge process by integrating GitHub repository rulesets into the required status check verification logic. It introduces helpers to fetch and match ruleset-based required contexts, supplementing classic branch protection and ensuring correct enforcement even when the classic API returns a 404. The test suite has been significantly updated to cover ruleset scenarios and now enforces that pending required checks block merges until they successfully conclude. Review feedback identifies opportunities to improve glob pattern matching, handle ruleset exclusion patterns, optimize API efficiency by avoiding N+1 queries, and deduplicate the final list of required contexts.
> 
> <details>
> <summary>I am having trouble creating individual review comments. Click here to see my feedback.</summary>
> 
> ### .agents/scripts/pulse-merge-process.sh (970-979)
> 
> ![high](https://www.gstatic.com/codereviewagent/high-priority.svg)
> 
> The glob detection pattern `refs/heads/*\**)` is overly restrictive. It only matches patterns that start with `refs/heads/` and contain an asterisk. However, GitHub rulesets also support simple globs like `*` (which matches all refs) or patterns that don't start with the full ref prefix. These would currently be ignored by the matching logic. Ensure that any fix is applied consistently to similar patterns throughout the file, and that the `$default_ref` is determined using a robust fallback mechanism.
> 
> ```suggestion
> 	case "$pattern" in
> 	*"*"*)
> 		# shellcheck disable=SC2254 # Intentionally treat ruleset branch globs as patterns.
> 		case "$default_ref" in
> 		$pattern)
> 			return 0
> 			;;
> 		esac
> 		;;
> 	esac
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. When applying a fix for a specific pattern, also fix other instances of the same pattern in the file for consistency.
> 2. To make git operations in scripts more robust, dynamically and explicitly determine the default branch instead of relying on the current checkout.
> </detai
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.14.86 automated scan.

--- a/.agents/scripts/pulse-merge-process.sh
+++ b/.agents/scripts/pulse-merge-process.sh
@@ -960,18 +960,30 @@ _ruleset_ref_matches_default_branch() {
 	local pattern="$1"
 	local default_branch="$2"
 	local default_ref="refs/heads/${default_branch}"
+	local branch_pattern="${pattern}"
 
 	case "$pattern" in
-	"~ALL" | "~DEFAULT_BRANCH" | "$default_ref")
+	"~ALL" | "~DEFAULT_BRANCH" | "$default_ref" | "$default_branch")
 		return 0
+		;;
+	esac
+	case "$pattern" in
+	refs/heads/*)
+		branch_pattern="${pattern#refs/heads/}"
 		;;
 	esac
 
 	case "$pattern" in
-	refs/heads/*\**)
+	*"*"*)
 		# shellcheck disable=SC2254 # Intentionally treat ruleset branch globs as patterns.
 		case "$default_ref" in
 		$pattern)
+			return 0
+			;;
+		esac
+		# shellcheck disable=SC2254 # Intentionally treat ruleset branch globs as patterns.
+		case "$default_branch" in
+		$branch_pattern)
 			return 0
 			;;
 		esac
@@ -1016,7 +1028,8 @@ _required_contexts_from_rulesets_for_default_branch() {
 		return 1
 	}
 
-	local id="" detail="" include_patterns="" pattern="" matches_default=0 contexts=""
+	local id="" detail="" include_patterns="" exclude_patterns="" pattern=""
+	local matches_default=0 excluded_default=0 contexts=""
 	while IFS= read -r id; do
 		[[ -n "$id" ]] || continue
 		detail=$(gh api "repos/${repo_slug}/rulesets/${id}" 2>/dev/null) || {
@@ -1030,6 +1043,11 @@ _required_contexts_from_rulesets_for_default_branch() {
 			rm -f "$contexts_tmp"
 			return 1
 		}
+		exclude_patterns=$(printf '%s' "$detail" | jq -r '.conditions.ref_name.exclude // [] | .[]' 2>/dev/null) || {
+			echo "[pulse-merge] _required_contexts_from_rulesets_for_default_branch: ruleset detail ${id} exclude parse failed for ${repo_slug} — caller will fail closed (GH#23019)" >>"$LOGFILE"
+			rm -f "$contexts_tmp"
+			return 1
+		}
 
 		matches_default=0
 		while IFS= read -r pattern; do
@@ -1040,6 +1058,16 @@ _required_contexts_from_rulesets_for_default_branch() {
 			fi
 		done <<<"$include_patterns"
 		[[ "$matches_default" -eq 1 ]] || continue
+
+		excluded_default=0
+		while IFS= read -r pattern; do
+			[[ -n "$pattern" ]] || continue
+			if _ruleset_ref_matches_default_branch "$pattern" "$default_branch"; then
+				excluded_default=1
+				break
+			fi
+		done <<<"$exclude_patterns"
+		[[ "$excluded_default" -eq 0 ]] || continue
 
 		contexts=$(printf '%s' "$detail" | jq -r '.rules[]? | select(.type == "required_status_checks") | (.parameters.required_status_checks // [])[]? | .context // empty' 2>/dev/null) || {
 			echo "[pulse-merge] _required_contexts_from_rulesets_for_default_branch: required-check parse failed for ruleset ${id} in ${repo_slug} — caller will fail closed (GH#23019)" >>"$LOGFILE"

--- a/.agents/scripts/tests/test-pulse-merge-phantom-pending-checks.sh
+++ b/.agents/scripts/tests/test-pulse-merge-phantom-pending-checks.sh
@@ -24,10 +24,13 @@
 #   5. no_required_contexts — no classic/ruleset contexts → return 0
 #   6. rulesets_required_on_404 — rulesets-only required check failing → return 1
 #   7. rulesets_required_on_404_pass — rulesets-only required check passing → return 0
-#   8. rulesets_non_matching_branch — non-default ruleset ignored → return 0
-#   9. default_branch_api_error — can't get default branch → return 1 (fail-closed)
-#   10. branch_protection_api_error — can't get branch protection → return 1 (fail-closed)
-#   11. checks_api_error — can't get REST check-runs → return 1 (fail-closed)
+#   8. rulesets_required_with_simple_glob — `*` ruleset applies → return 1
+#   9. rulesets_required_with_plain_branch — plain branch pattern applies → return 1
+#   10. rulesets_excludes_default_branch — exclusion wins → return 0
+#   11. rulesets_non_matching_branch — non-default ruleset ignored → return 0
+#   12. default_branch_api_error — can't get default branch → return 1 (fail-closed)
+#   13. branch_protection_api_error — can't get branch protection → return 1 (fail-closed)
+#   14. checks_api_error — can't get REST check-runs → return 1 (fail-closed)
 #
 # Strategy: source shared-gh-wrappers-checks.sh for `gh_pr_check_runs_rest`,
 # extract _check_required_checks_passing from pulse-merge-process.sh, eval it,
@@ -109,8 +112,10 @@ setup_test_env() {
 # Mode env vars:
 #   MOCK_REPO_MODE     — ok | error
 #   MOCK_BP_MODE       — three_required | no_required | not_found | error
-#   MOCK_RULESETS_MODE — none | active_required | active_required_other_branch |
-#                        error | detail_error
+#   MOCK_RULESETS_MODE — none | active_required | active_required_star |
+#                        active_required_plain_branch |
+#                        active_required_excluded_default |
+#                        active_required_other_branch | error | detail_error
 #   MOCK_ROLLUP_MODE   — all_required_pass | phantom_pending | one_required_failing |
 #                        required_absent | error
 printf '%s\n' "$*" >> "${GH_CALL_LOG}"
@@ -143,6 +148,18 @@ if [[ "$1" == "api" && "$2" == repos/* && "$*" == *"/rulesets/"* ]]; then
 		apply_jq '{"conditions":{"ref_name":{"include":["~DEFAULT_BRANCH"]}},"rules":[{"type":"required_status_checks","parameters":{"required_status_checks":[{"context":"review-bot-gate"}]}}]}' "$@"
 		exit 0
 		;;
+	active_required_star)
+		apply_jq '{"conditions":{"ref_name":{"include":["*"]}},"rules":[{"type":"required_status_checks","parameters":{"required_status_checks":[{"context":"review-bot-gate"}]}}]}' "$@"
+		exit 0
+		;;
+	active_required_plain_branch)
+		apply_jq '{"conditions":{"ref_name":{"include":["main"]}},"rules":[{"type":"required_status_checks","parameters":{"required_status_checks":[{"context":"review-bot-gate"}]}}]}' "$@"
+		exit 0
+		;;
+	active_required_excluded_default)
+		apply_jq '{"conditions":{"ref_name":{"include":["*"],"exclude":["main"]}},"rules":[{"type":"required_status_checks","parameters":{"required_status_checks":[{"context":"review-bot-gate"}]}}]}' "$@"
+		exit 0
+		;;
 	active_required_other_branch)
 		apply_jq '{"conditions":{"ref_name":{"include":["refs/heads/release"]}},"rules":[{"type":"required_status_checks","parameters":{"required_status_checks":[{"context":"review-bot-gate"}]}}]}' "$@"
 		exit 0
@@ -165,7 +182,8 @@ if [[ "$1" == "api" && "$2" == repos/* && "$*" == *"/rulesets"* ]]; then
 		apply_jq '[]' "$@"
 		exit 0
 		;;
-	active_required | active_required_other_branch | detail_error)
+	active_required | active_required_star | active_required_plain_branch | \
+		active_required_excluded_default | active_required_other_branch | detail_error)
 		apply_jq '[{"id":101,"enforcement":"active"}]' "$@"
 		exit 0
 		;;
@@ -504,6 +522,46 @@ test_rulesets_required_on_branch_protection_404_pass() {
 	return 0
 }
 
+test_rulesets_required_with_simple_glob() {
+	reset_logs
+	export MOCK_REPO_MODE="ok"
+	export MOCK_BP_MODE="not_found"
+	export MOCK_RULESETS_MODE="active_required_star"
+	export MOCK_ROLLUP_MODE="one_required_failing"
+	assert_returns 1 "branch protection 404 + simple '*' ruleset glob → blocked"
+	assert_log_contains "active rulesets require contexts" \
+		"rulesets_star_glob: ruleset context logged"
+	assert_log_contains "required context(s) not passing" \
+		"rulesets_star_glob: block message logged"
+	return 0
+}
+
+test_rulesets_required_with_plain_branch() {
+	reset_logs
+	export MOCK_REPO_MODE="ok"
+	export MOCK_BP_MODE="not_found"
+	export MOCK_RULESETS_MODE="active_required_plain_branch"
+	export MOCK_ROLLUP_MODE="one_required_failing"
+	assert_returns 1 "branch protection 404 + plain default-branch ruleset → blocked"
+	assert_log_contains "active rulesets require contexts" \
+		"rulesets_plain_branch: ruleset context logged"
+	assert_log_contains "required context(s) not passing" \
+		"rulesets_plain_branch: block message logged"
+	return 0
+}
+
+test_rulesets_excludes_default_branch() {
+	reset_logs
+	export MOCK_REPO_MODE="ok"
+	export MOCK_BP_MODE="not_found"
+	export MOCK_RULESETS_MODE="active_required_excluded_default"
+	export MOCK_ROLLUP_MODE="one_required_failing"
+	assert_returns 0 "branch protection 404 + ruleset excluding default branch → allowed"
+	assert_log_contains "no classic branch protection or required ruleset contexts" \
+		"rulesets_excluded_default: empty-context message logged"
+	return 0
+}
+
 test_rulesets_non_matching_branch_ignored() {
 	reset_logs
 	export MOCK_REPO_MODE="ok"
@@ -567,6 +625,9 @@ main() {
 	test_no_required_contexts
 	test_rulesets_required_on_branch_protection_404_block
 	test_rulesets_required_on_branch_protection_404_pass
+	test_rulesets_required_with_simple_glob
+	test_rulesets_required_with_plain_branch
+	test_rulesets_excludes_default_branch
 	test_rulesets_non_matching_branch_ignored
 	test_default_branch_api_error
 	test_branch_protection_api_error

